"Building multiple Progressive Web Apps on the same domain" article. (#5389)
GoogleChrome/web.dev: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md

imposes restrictions on "origins", so we'll mostly refer to the term throughout
the article. Nevertheless, we'll use "domains" or "subdomains" from time to
time, to describe the technique being used, in order to create the different
"origins".

## The case for multiple, related PWAs

In some cases, you might want to build independent apps, but still identify them
as belonging to the same organization or "brand". Reusing the same **domain
name** is a good way of establishing that relationship. For example:

- An ecommerce site wants to create a standalone experience to let sellers
  manage their inventory, while making sure they understand that it belongs to
  the main website where the users buy products.
- A sports news site wants to build a specific app for a major sporting event,
  to let users receive stats about their favorite competitions via
  notifications, and install it as a Progressive Web App, while making sure that
  users recognize it as an app built by the news company.
- A company wants to build separate chat, mail and calendar apps and wants them
  to work as individual apps, tied to the company's name.

<figure class="w-figure">

>   {% Img src="image/26V1DWN36MZr3mUo8ChSBlCpzp43/XEwhGkkV9Bz7Z2KEJWjm.png", alt="Avoid using different origins for site sections of the same site when trying to build a unified Progresive Web App.", width="800", height="475" %}

Author: Hi @jpmedley! That's what I wanted to say. There are site that have the home page at https://www.example.com (origin 1) and the checkout page at https://checkout.example.com (origin 2), and the login page at https://login.example.com (origin 3). This is actually a unique site, that, for some reason has been divided into multiple origins by using subdomains. If the developer wants to provide a single PWA experience across the site, that will have many challenges. That's why that alternative is not recommended.

Please, let me know if that's clear and if there's a better way of explaining it in the alternative image text.

<figure class="w-figure">
  {% Img src="image/26V1DWN36MZr3mUo8ChSBlCpzp43/FfuseK64y4PRqXYEGEt9.png", alt="ALT_TEXT_HERE", width="800", height="421" %}
  <figcaption class="w-figcaption">Building different PWAs in distinct origins, by using subdomains is a good practice.</figcaption>
</figure>

## Using the same origin

The second approach is building the different PWAs on the same origin. This
includes the following scenarios:

### Non-overlapping paths

Multiple PWAs or conceptual "web apps", hosted on the same origin, with
non-overlapping paths. For example:

- `https://example.com/app1/`
- `https://example.com/app2/`

### Overlapping/nested paths

Multiple PWAs on the same origin, one of whose scope is nested inside the other:

- `https://example.com/` (the "outer app")
- `https://example.com/app/` (the "inner app")

The service worker API and manifest format allow you to do either of the above,
using path-level scoping. However, in both cases, using the same origin presents
many problems and limitations, the root of which stems from the fact that the
browser won't fully consider these to be distinct "apps", therefore **this

> approach is discouraged**.

Author: I understand that in most cases we tend to talk about the "positive" case and avoid talking too much about the "negative" one and just say that it's not recommended. 
In this particular article, talking about the non-recommended use case is important because is a very common mistake developers do, and they don't really understand why it's a mistake.

Let me know if that makes sense.

cc // @mgiuca

<figure class="w-figure">
  {% Img src="image/26V1DWN36MZr3mUo8ChSBlCpzp43/6HX5zEb58sEEWQoJt82m.png", alt="ALT_TEXT_HERE", width="800", height="420" %}
  <figcaption class="w-figcaption">Using paths (overlapping or not) to provide two independent PWAs (“app1”, “app2”) under the same origin is discouraged.</figcaption>
</figure>

In the next section, we analyze these challenges in more detail, and what can be
done, if using separate origins is not an option.

## Challenges for multiple, same-origin PWAs

> Here are some practical issues common to both "same-origin" approaches:

Reviewer: ```suggestion
Here are some practical issues common to both same-origin approaches:
```

- **Storage:** Cookies, local storage, and all forms of device-local storage are
  shared between apps. For that reason, if the user decides to wipe local data

>   for one app, it will wipe all the data from the origin; there's no way of
>   indicating to do this for a "single app". Note that Chrome and some other

Reviewer: ```suggestion
  for one app, it will wipe all the data from the origin; there's no way 
 to do this for a single app. Note that Chrome and some other
```

  browsers will actively prompt users to wipe local data when uninstalling one
  of the apps, and this will affect data for the other apps on the origin as

>   well!

Reviewer: ```suggestion
  well.
```

- **Permissions:** Permissions are tied to the origin. That means if the user
  grants a permission to one app, it will apply to all apps on that origin
  simultaneously. That may sound like a good thing (not having to ask for a
  permission multiple times), but remember: if the user blocks permission to one
  app, it will prevent the others from requesting that permission or using that

>   feature!

Reviewer: ```suggestion
  feature.
```

- **User settings:** Settings are also set per-origin. For example, if two apps
  have different font sizes, and the user wants to adjust the zoom in only one
  of them to compensate for it, they won't be able to do it without applying the
  setting to the other apps as well.

These challenges make it difficult to encourage this approach. Nevertheless, if

> you can't use a separate origin (e.g. a subdomain), as discussed in the "Using
> separate origins" section, from the two "same-origin" options we presented,

Reviewer: ```suggestion
you can't use a separate origin (e.g. a subdomain), as discussed in the [Using
separate origins](#using-separate-origins) section, from the two same-origin options we presented,
```

> using **non-overlapping paths** is strongly recommended, over
> **overlapping/nested paths**.

Reviewer: ```suggestion
using non-overlapping paths is strongly recommended, over
overlapping/nested paths.
```

As mentioned, the challenges discussed in this section are common to both

> "same-origin" approaches. In the next section we'll go deeper into the details

Reviewer: ```suggestion
same-origin approaches. In the next section we'll go deeper into the details
```

> of why using **overlapping/nested paths** is the least recommended strategy.

Reviewer: ```suggestion
of why using overlapping/nested paths is the least recommended strategy.
```

## Additional challenges for overlapping/nested paths

The additional issue with overlapping/nested paths approach (where

> `https://example.com/` is the "outer app" and  `https://example.com/app/` is the

Reviewer: ```suggestion
`https://example.com/` is the outer app and  `https://example.com/app/` is the
```

> "inner app"), is that all URLs in the "inner app" will actually be considered to

Reviewer: ```suggestion
inner app), is that all URLs in the inner app will actually be considered
```

> be part of both the "outer app" and the "inner app".

Reviewer: ```suggestion
part of both the outer app and the inner app.
```

In practice this presents the following issues:

> - **Installation Promotion:** If the user visits the "inner app" (for example,

Reviewer: ```suggestion
- **Installation Promotion:** If the user visits the inner app (for example,
```

>   in a web browser), when the "outer app" is already installed in the user's

Reviewer: ```suggestion
  in a web browser), when the outer app is already installed in the user's
```

  device, the browser won't show the install promotional banners, and the
  [BeforeInstallPrompt event](https://web.dev/customize-install/) won't be
  triggered. The reason is that the browser will check and see whether the
  current page belongs to an app that's already installed, and it will conclude

>   that it is. The workarounds for this is to install the "inner app" manually

Reviewer: ```suggestion
  that it is. The workaround for this is to install the inner app manually
```

>   (via "Create Shortcut" browser menu option), or to install the "inner app"

Reviewer: ```suggestion
  (via "Create Shortcut" browser menu option), or to install the inner app
```

>   first before the "outer app".

Reviewer: ```suggestion
  first, before the outer app.
```

- **[Notification](https://developer.mozilla.org/en-US/docs/Web/API/notification)

>   and [Badging API](https://web.dev/badging-api/)**: If the "outer app" is

Reviewer: ```suggestion
  and the [Badging API](https://web.dev/badging-api/)**: If the outer app is
```

>   installed but the "inner app" is not, notifications and badges coming from the

Reviewer: ```suggestion
  installed but the inner app is not, notifications and badges coming from the
```

>   "inner app" will be erroneously attributed to the "outer app" (which is the

Reviewer: ```suggestion
  inner app will be erroneously attributed to the outer app (which is the
```

  nearest enclosing scope of an installed app). This feature works properly in
  the case where both apps are installed on the user's device.
- **[Link
  Capturing](https://github.com/WICG/sw-launch/blob/main/declarative_link_capturing.md)**:

>   The "outer app" may capture URLs that belong to the "inner app". This is

Reviewer: ```suggestion
  The outer app may capture URLs that belong to the inner app. This is
```

>   especially likely if the "outer app" is installed but the "inner app" isn't.

Reviewer: ```suggestion
  especially likely if the outer app is installed but the inner app isn't.
```

>   Similarly, links within the "outer app" that link to the "inner app" will not

Reviewer: ```suggestion
  Similarly, links within the outer app that link to the inner app will not
```

>   link capture into the "inner app", since they are considered to be within the

Reviewer: ```suggestion
  link capture into the inner app, since they are considered to be within the
```

>   "outer app"'s scope. Additionally, on Chrome OS / Android, if these apps are

Reviewer: ```suggestion
  outer app's scope. Additionally, on Chrome OS and Android, if these apps are
```

  added to the Play Store (as [Trusted Web
  Activities](https://developer.chrome.com/docs/android/trusted-web-activity/overview/)),

>   the "outer app" will capture all links. Even if the "inner app" is installed,

Reviewer: ```suggestion
  the outer app will capture all links. Even if the inner app is installed,
```

>   the OS will still offer the user the choice of opening them in the "outer
>   app."

Reviewer: ```suggestion
  the OS will still offer the user the choice of opening them in the outer
  app.
```

## Conclusion

In this article we looked at different ways in which developers can build

> multiple Progressive Web Apps related to each other within the same **domain
> name**.

Reviewer: ```suggestion
multiple Progressive Web Apps related to each other within the same **domain**.
```

Reviewer: This is to be consistent with what you've done elsewhere.

In summary, we strongly recommend using a different origin (e.g. by using
subdomains) to host independent PWAs. Hosting them in the same origin presents

> many challenges, mainly due to the fact that the browser won't fully consider

Reviewer: ```suggestion
many challenges, mainly because the browser won't fully consider
```

> these to be distinct "apps".

Reviewer: ```suggestion
these to be distinct apps.
```

- Separate origins: **Recommended**
- Same origin, non-overlapping paths: **Not recommended**
- Same origin, overlapping/nested paths: **Strongly not recommended**

If it's not possible to use different origins, using non-overlapping paths (e.g.
`https://example.com/app1/` and `https://example.com/app2/` it's strongly
recommended over using overlapping or nested paths, like `https://example.com/`

> (for the "outer app") and `https://example.com/app/` (for the "inner app").

Reviewer: ```suggestion
(for the outer app) and `https://example.com/app/` (for the inner app).
```
